File: src/lib/integrations/github-webhook.ts
import crypto from "crypto";
import { db } from "@/lib/db";
import { processGitHubEvent } from "@/lib/agent/github-processor";

export function verifyGitHubSignature(
  payload: string,
  signature: string | null,
  secret: string
): boolean {
  if (!signature) return false;

  const expected =
    "sha256=" +
    crypto.createHmac("sha256", secret).update(payload).digest("hex");

  try {
    return crypto.timingSafeEqual(
      Buffer.from(signature),
      Buffer.from(expected)
    );
  } catch {
    return false;
  }
}

export async function handleGitHubWebhook(
  deliveryId: string,
  event: string,
  payload: Record<string, unknown>
) {
  const existing = await db.webhookDelivery.findUnique({
    where: { id: deliveryId },
  });
  if (existing) return;

  await db.webhookDelivery.create({
    data: { id: deliveryId, provider: "github" },
  });

  await processGitHubEvent(event, payload);
}

File: backend/app/services/integrations/github_webhook.py
import hashlib
import hmac

from sqlalchemy import select

from app.database import AsyncSessionLocal
from app.models import WebhookDelivery
from app.services.agent.github_processor import process_github_event


def verify_github_signature(payload: str, signature: str | None, secret: str) -> bool:
    if not signature:
        return False

    expected = "sha256=" + hmac.new(
        secret.encode(), payload.encode(), hashlib.sha256
    ).hexdigest()

    try:
        return hmac.compare_digest(signature, expected)
    except Exception:
        return False


async def handle_github_webhook(
    delivery_id: str,
    event: str,
    payload: dict,
) -> None:
    async with AsyncSessionLocal() as db:
        result = await db.execute(
            select(WebhookDelivery).where(WebhookDelivery.id == delivery_id)
        )
        if result.scalar_one_or_none():
            return

        db.add(WebhookDelivery(id=delivery_id, provider="github"))
        await db.commit()

    await process_github_event(event, payload)
